store/etcd/types_transformation.go

```diff
@@ -0,0 +1,182 @@
+package etcd
+
+import (
+	"fmt"
+
+	"github.com/heetch/regula"
+	"github.com/heetch/regula/rule"
+	pb "github.com/heetch/regula/store/etcd/proto"
+)
+
+// toProtobufRuleset creates a protobuf Ruleset from a regula.Ruleset.
+func toProtobufRuleset(rs *regula.Ruleset) *pb.Ruleset {
+	pbrs := &pb.Ruleset{
+		Type: rs.Type,
+	}
+
+	for _, r := range rs.Rules {
+		pbr := &pb.Rule{
+			Expr:   toProtobufExpr(r.Expr),
+			Result: toProtobufValue(r.Result),
+		}
+		pbrs.Rules = append(pbrs.Rules, pbr)
+	}
+
+	return pbrs
+}
+
+// toProtobufExpr creates a protobuf Expr from a rule.Expr.
+func toProtobufExpr(expr rule.Expr) *pb.Expr {
+	switch e := expr.(type) {
+	case *rule.Value:
+		v := &pb.Value{
+			Type: e.Type,
+			Kind: e.Kind,
+			Data: e.Data,
+		}
+
+		return &pb.Expr{
+			Expr: &pb.Expr_Value{Value: v},
+		}
+	case *rule.Param:
+		p := &pb.Param{
+			Kind: e.Kind,
+			Type: e.Type,
+			Name: e.Name,
+		}
+
+		return &pb.Expr{
+			Expr: &pb.Expr_Param{Param: p},
+		}
+	}
+
+	var (
+		ope rule.Operander
+		ok  bool
+	)
+	if ope, ok = expr.(rule.Operander); !ok {
+		// there is something very weird, a rule.Expr which is not a rule.Value nor a rule.Param nor a rule.Operander
+		// let's panic...
+		panic(fmt.Sprintf("cannot create a pb.Expr - unexpected type: %T", expr))
+	}
+
+	o := &pb.Operator{
+		Kind: expr.Contract().OpCode,
+	}
+	for _, op := range ope.Operands() {
+		o.Operands = append(o.Operands, toProtobufExpr(op))
+	}
+
+	return &pb.Expr{
+		Expr: &pb.Expr_Operator{Operator: o},
+	}
+}
+
+// toProtobufValue creates a protobuf Value from a rule.Value.
+func toProtobufValue(val *rule.Value) *pb.Value {
+	return &pb.Value{
+		Kind: val.Kind,
+		Type: val.Type,
+		Data: val.Data,
+	}
+}
+
+// toProtobufSignature creates a protobuf Signature from a regula.Signature.
+func toProtobufSignature(sig *regula.Signature) *pb.Signature {
+	pbsig := &pb.Signature{
+		ReturnType: sig.ReturnType,
+		ParamTypes: make(map[string]string),
+	}
+
+	for k, v := range sig.ParamTypes {
+		pbsig.ParamTypes[k] = v
+	}
+
+	return pbsig
+}
+
+// fromProtobufRuleset creates a regula.Ruleset from a protobuf Ruleset.
+func fromProtobufRuleset(pbrs *pb.Ruleset) *regula.Ruleset {
+	rs := &regula.Ruleset{
+		Type: pbrs.Type,
+	}
+
+	for _, r := range pbrs.Rules {
+		rr := &rule.Rule{
+			Expr:   fromProtobufExpr(r.Expr),
+			Result: fromProtobufValue(r.Result),
+		}
+		rs.Rules = append(rs.Rules, rr)
+	}
+
+	return rs
+}
+
+// fromProtobufExpr creates a rule.Expr from a protobuf Expr.
+func fromProtobufExpr(expr *pb.Expr) rule.Expr {
+	switch e := expr.Expr.(type) {
+	case *pb.Expr_Value:
+		return &rule.Value{
```

Comment:
:book: A trick since Go 1.9 IIRC, when two structs have the same fields, you can simply cast one into the other:

```go
// not sure this works but it's something like that
return &(rule.Value)(e.Value)
```

Reply:
Nice one but it can't work here because `e.Value` is a protobuf type which has more fields (the `XXX_...` ones):
```go
type Value struct {
	Kind                 string   `protobuf:"bytes,1,opt,name=kind,proto3" json:"kind,omitempty"`
	Type                 string   `protobuf:"bytes,2,opt,name=type,proto3" json:"type,omitempty"`
	Data                 string   `protobuf:"bytes,3,opt,name=data,proto3" json:"data,omitempty"`
	XXX_NoUnkeyedLiteral struct{} `json:"-"`
	XXX_unrecognized     []byte   `json:"-"`
	XXX_sizecache        int32    `json:"-"`
}
```

Reply:
Indeed it can't work here :+1: